x-pack/plugins/lens/public/editor_frame_plugin/editor_frame/suggestion_helpers.ts

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { DatasourcePublicAPI, TableColumn, Visualization, DatasourceSuggestion } from '../../types';
+import { Action } from './state_management';
+
+export interface Suggestion {
+  visualizationId: string;
+  datasourceState: unknown;
+  score: number;
+  title: string;
+  state: unknown;
+}
+
+/**
+ * This function takes a list of available data tables and a list of visualization
+ * extensions and creates a ranked list of suggestions which contain a pair of a data table
+ * and a visualization.
+ *
+ * Each suggestion represents a valid state of the editor and can be applied by creating an
+ * action with `toSwitchAction` and dispatching it
+ */
+export function getSuggestions(
+  datasourceTableSuggestions: DatasourceSuggestion[],
+  visualizationMap: Record<string, Visualization>,
+  activeVisualizationId: string | null,
+  visualizationState: unknown,
+  datasourcePublicAPI: DatasourcePublicAPI
+): Suggestion[] {
+  const datasourceTableMetas: Record<string, TableColumn[]> = {};
+  datasourceTableSuggestions.map(({ tableColumns }, datasourceSuggestionId) => {
+    datasourceTableMetas[datasourceSuggestionId] = tableColumns;
+  });
+
+  return (
+    Object.entries(visualizationMap)
+      .map(([visualizationId, visualization]) => {
+        return visualization
+          .getSuggestions({
+            tableColumns: datasourceTableMetas,
+            state: visualizationId === activeVisualizationId ? visualizationState : undefined,
+          })
+          .map(({ datasourceSuggestionId, ...suggestion }) => ({
+            ...suggestion,
+            visualizationId,
+            datasourceState: datasourceTableSuggestions[datasourceSuggestionId].state,
+          }));
+      })
+      // TODO why is flatMap not available here?
+      .reduce((globalList, currentList) => [...globalList, ...currentList], [])
```

Comment:
flatMap seems to be there for me

Reply:
The types are there, but it fails in jest